internal/controllers/rollout/integration_test.go

```diff
@@ -75,6 +75,81 @@ func TestSynthesizerRollout(t *testing.T) {
 	})
 }
 
+// TestRolloutIgnoreSideEffects proves that synthesizer changes are not rolled out to compositions which are ignoring side effects.
+func TestRolloutIgnoreSideEffects(t *testing.T) {
+	ctx := testutil.NewContext(t)
+	mgr := testutil.NewManager(t)
+	cli := mgr.GetClient()
+
+	require.NoError(t, flowcontrol.NewSynthesisConcurrencyLimiter(mgr.Manager, 10, 0))
+	require.NoError(t, NewSynthesizerController(mgr.Manager))
+	require.NoError(t, NewController(mgr.Manager, time.Millisecond*10))
+	require.NoError(t, synthesis.NewPodLifecycleController(mgr.Manager, testSynthesisConfig))
+	testutil.WithFakeExecutor(t, mgr, func(ctx context.Context, s *apiv1.Synthesizer, input *krmv1.ResourceList) (*krmv1.ResourceList, error) {
+		output := &krmv1.ResourceList{}
+		return output, nil
+	})
+	mgr.Start(t)
+
+	syn := &apiv1.Synthesizer{}
+	syn.Name = "test-syn"
+	syn.Spec.Image = "test-syn-image"
+	require.NoError(t, cli.Create(ctx, syn))
+
+	comp := &apiv1.Composition{}
+	comp.Name = "test-comp"
+	comp.Namespace = "default"
+	comp.Spec.Synthesizer.Name = syn.Name
+	comp.Spec.IgnoreSideEffects = true
+	require.NoError(t, cli.Create(ctx, comp))
+
+	// Initial creation.
+	testutil.Eventually(t, func() bool {
+		require.NoError(t, client.IgnoreNotFound(cli.Get(ctx, client.ObjectKeyFromObject(comp), comp)))
+		return comp.Status.CurrentSynthesis != nil && comp.Status.CurrentSynthesis.Synthesized != nil
+	})
+
+	// Update the synthesizer while ignoring side effects.
+	err := retry.RetryOnConflict(testutil.Backoff, func() error {
+		if err := cli.Get(ctx, client.ObjectKeyFromObject(syn), syn); err != nil {
+			return err
+		}
+		syn.Spec.Image = "updated-image"
+		return cli.Update(ctx, syn)
+	})
+	require.NoError(t, err)
+	// Give some time to the controller to process the change.
+	time.Sleep(time.Second)
```

Comment:
wait short time might cause the unit test to be flaky, can we use testuitl.Eventually to check the syn.metadata.resourceVersion is changed after updating the spec.image to syn?

Reply:
The issue here is that we're waiting for it to _not change_ so we can't use eventually. The only way that I can think of doing this is by somehow hooking into controller runtime and ensure that this write has been processed, but even then I don't know if it can be completely trusted. It's a good idea to explore other ways of testing negative cases though.

Reply:
i see, thanks for the explanation 👍